res/res_pjsip_stir_shaken.c

```diff
@@ -181,222 +230,220 @@ static void stir_shaken_inv_end_session(struct ast_sip_session *session, pjsip_r
+		SCOPE_EXIT_RTN_VALUE(1, "%s: Unable to create context.  Call terminated\n",
+			session_name);
 	}
 
-	identity_hdr_val = ast_sip_rdata_get_header_value(rdata, identity_str);
+	identity_hdr_val = ast_sip_rdata_get_header_value(rdata, identity_hdr_str);
 	if (ast_strlen_zero(identity_hdr_val)) {
-		ast_stir_shaken_add_verification(chan, caller_id, "", AST_STIR_SHAKEN_VERIFY_NOT_PRESENT);
-		return 0;
+		p_rc = process_failure(ctx, caller_id, session, rdata,
+			AST_STIR_SHAKEN_VS_NO_IDENTITY_HDR);
+		if (p_rc == PROCESS_FAILURE_CONTINUE) {
+			SCOPE_EXIT_RTN_VALUE(0, "%s: No Identity header found.  Call continuing\n",
+				session_name);
+		}
+		SCOPE_EXIT_LOG_RTN_VALUE(1, LOG_ERROR, "%s: No Identity header found.  Call terminated\n",
+			session_name);
 	}
 
-	encoded_val = strtok_r(identity_hdr_val, ".", &identity_hdr_val);
-	header = ast_base64url_decode_string(encoded_val);
-	if (ast_strlen_zero(header)) {
-		ast_debug(3, "STIR/SHAKEN INVITE for %s is missing header\n",
-			ast_sorcery_object_get_id(session->endpoint));
-		stir_shaken_inv_end_session(session, rdata, AST_STIR_SHAKEN_RESPONSE_CODE_BAD_IDENTITY_INFO, bad_identity_info_str);
-		return 1;
+	vs_rc = ast_stir_shaken_vs_ctx_add_identity_hdr(ctx, identity_hdr_val);
+	if (vs_rc != AST_STIR_SHAKEN_VS_SUCCESS) {
+		reject_incoming_call(session, 500);
+		SCOPE_EXIT_LOG_RTN_VALUE(1, LOG_ERROR, "%s: Unable to add Identity header.  Call terminated.\n",
+			session_name);
 	}
 
-	encoded_val = strtok_r(identity_hdr_val, ".", &identity_hdr_val);
-	payload = ast_base64url_decode_string(encoded_val);
-	if (ast_strlen_zero(payload)) {
-		ast_debug(3, "STIR/SHAKEN INVITE for %s is missing payload\n",
-			ast_sorcery_object_get_id(session->endpoint));
-		stir_shaken_inv_end_session(session, rdata, AST_STIR_SHAKEN_RESPONSE_CODE_BAD_IDENTITY_INFO, bad_identity_info_str);
-		return 1;
+	date_hdr_val = ast_sip_rdata_get_header_value(rdata, date_hdr_str);
+	if (ast_strlen_zero(date_hdr_val)) {
+		p_rc = process_failure(ctx, caller_id, session, rdata,
+			AST_STIR_SHAKEN_VS_NO_DATE_HDR);
+		if (p_rc == PROCESS_FAILURE_CONTINUE) {
+			SCOPE_EXIT_RTN_VALUE(0, "%s: No Date header found.  Call continuing\n",
+				session_name);
+		}
+		SCOPE_EXIT_LOG_RTN_VALUE(1, LOG_ERROR, "%s: No Date header found.  Call terminated\n",
+			session_name);
 	}
 
-	/* It's fine to leave the signature encoded */
-	signature = strtok_r(identity_hdr_val, ";", &identity_hdr_val);
-	if (ast_strlen_zero(signature)) {
-		ast_debug(3, "STIR/SHAKEN INVITE for %s is missing signature\n",
-			ast_sorcery_object_get_id(session->endpoint));
-		stir_shaken_inv_end_session(session, rdata, AST_STIR_SHAKEN_RESPONSE_CODE_BAD_IDENTITY_INFO, bad_identity_info_str);
-		return 1;
+	ast_stir_shaken_vs_ctx_add_date_hdr(ctx, date_hdr_val);
+	if (vs_rc != AST_STIR_SHAKEN_VS_SUCCESS) {
+		reject_incoming_call(session, 500);
+		SCOPE_EXIT_LOG_RTN_VALUE(1, LOG_ERROR, "%s: Unable to add Date header.  Call terminated.\n",
+			session_name);
 	}
```

Comment:
If this fails, should it also check to see if it should continue or not?

Reply:
I terminated here because the only failure possible in ast_stir_shaken_vs_ctx_add_date_hdr is a failure to allocate memory so the call probably isn't going to succeed anyway.